kong/router/transform.lua

```diff
@@ -733,42 +733,36 @@ local function split_routes_and_services_by_path(routes_and_services)
       goto continue
     end
 
-    -- make sure that route_and_service contains
-    -- only the two expected entries, route and service
+    local grouped_paths_count, grouped_paths = group_by_regex_or_length(original_paths)
+    if grouped_paths_count == 1 then
+      goto continue
+    end
+
+    -- make sure that route_and_service contains only
+    -- the two expected entries, route and service
     local nkeys = tb_nkeys(route_and_service)
     assert(nkeys == 1 or nkeys == 2)
 
     local original_route_id = original_route.id
     local original_service = route_and_service.service
 
-    -- `grouped_paths` is a hash table, like {true={'regex'}, 2={'/a'}, 3={'/aa'},}
-    local grouped_paths = group_by(original_paths, sort_by_regex_or_length)
-
-    local is_first = true
-    for key, paths in pairs(grouped_paths) do
+    for j = 1, grouped_paths_count do
       local route = shallow_copy(original_route)
-
       -- create a new route from the original route
       route.original_route = original_route
-      route.paths = paths
-      route.id = uuid_generator(original_route_id .. "#" .. tostring(key))
+      route.paths = grouped_paths[j]
+      route.id = uuid_generator(original_route_id .. "#" .. j)
 
-      local cloned_route_and_service = {
+      local index = j == 1 and i or count + 1
```

Comment:
We should add some comments for this appending logic, it is a little hard to understand.